pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,520 @@
+			manager.deleteEgressMap(config, podData)
+		}
+		delete(manager.policyPods, id)
+	}
+	delete(manager.podDataStore, id)
+}
+
+func (manager *Manager) getPodMetadata(pod *slimcorev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := k8s.ServiceID{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		log.Warn("Failed in getting namespace")
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (m *Manager) getNamespaceForPod(pod *slimcorev1.Pod) (*slimcorev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := m.storeGetter.GetStore("namespace")
+	nsItem, exists, err := nsStore.GetByKey(pod.GetNamespace())
+	if err != nil {
+		log.WithError(err).Warn("Failed to get namespace")
+		return nil, err
+	}
+	if !exists {
+		log.WithFields(logrus.Fields{"nsName": nsName}).Warn("Namespace does not exist")
+		return nil, nil
+	}
+	return nsItem.(*slimcorev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	return nil
+}
+
+// upsertPolicyPod updates or insert to pod policy mapping for given policy config and pods,
+// it also upserts the pod data catch and egress map to keep in sync
+func (manager *Manager) upsertPolicyPod(config *EgressPolicyConfig, pods ...*podMetadata) {
+	for _, pod := range pods {
+		podIp := net.ParseIP(pod.ips[0])
+		if podPolicies, ok := manager.policyPods[pod.id]; ok {
+			newPodPolicy := true
+			for _, poID := range podPolicies {
+				// Existing pod policy update, noop for this policy
+				if poID == config.id {
+					newPodPolicy = false
+					break
+				}
+			}
+			if newPodPolicy {
+				// Pod selected by a new policy
+				manager.policyPods[pod.id] = append(manager.policyPods[pod.id], config.id)
+				manager.updateEgressMap(podIp, config)
+				manager.podDataStore[pod.id] = pod
+			}
+		} else {
+			// Pod selected by a policy for the first time
+			pp := []policyID{config.id}
+			manager.policyPods[pod.id] = pp
+			manager.updateEgressMap(podIp, config)
+			manager.podDataStore[pod.id] = pod
+		}
+	}
+}
+
+func (manager *Manager) updateEgressMap(sip net.IP, config *EgressPolicyConfig) error {
+	var err error
+	for _, dstCIDR := range config.dstCIDRs {
+		//TODO validate fields before updating the map
+		key := egressmap.NewKey(sip, dstCIDR.IP, dstCIDR.Mask)
+		value := &egressmap.EgressInfo4{}
+		copy(value.TunnelEndpoint[:], config.egressIp)
+		copy(value.EgressIP[:], config.egressIp)
```

Comment:
Why are `TunnelEndpoint` and `EgressIP` both set to `egressIP`?

Reply:
This is by design of the egress NAT feature, where we expect the egress IP to be assigned to the designated gateway node. So the egress ip input from the user is then going to be used for both destination ip for redirecting and then for SNATing at the gateway node.

Reply:
Ah, I thought the egress gateway would have several IP addresses assigned and the map would be used to select the IP address to use for SNATing. If it's going to have a single IP address like today, I don't really see the need for this `EgressIP` field in the map. Traffic leaving the gateway node will already be SNATed with the node's IP today.

Reply:
@pchaigno i think what you are saying is also true. There are multiple ips assigned to the gateway nodes. And for each egress policy it's up to the user to decide what ip to use for SNAT. What i meant is that the ip is also used to forward the traffic to the gateway.

Reply:
> What i meant is that the ip is also used to forward the traffic to the gateway.

That will only work if the egress IP is in the node CIDR. If it's *not* in the node CIDR, the encapsulated packets won't be correctly routed.

Also, if we're going for TunnelEndpoint  == EgressIP, why do we have two fields in the BPF map?

Reply:
> Also, if we're going for TunnelEndpoint == EgressIP, why do we have two fields in the BPF map?

I think we are reserving that probability even though it's not utilized at the moment. The ebpf program consume the two fields in different ways. @anfernee could you chime in on this?

Reply:
@pchaigno you are right. the current approach doesn't deal with multiNIC. The egress IPs have to be programmed on the same NIC as the node IP. The benefit is that when the egress IP floats from one node to another, we don't need to change the map everywhere to point it to the new node. Those 2 IPs serves for different purpose, even if they are the same in this implementation, it doesn't mean it's true in future, so we use 2 fields instead of one.

Reply:
Ok, makes sense :+1: 

@MasterZ40 Could you add a comment to explain that assumption in the code?

Reply:
Updated. Thx!
Resolving.